sdk/lib/atl/atlconv.h

```diff
@@ -0,0 +1,317 @@
+        Init(psz);
+    }
+
+    CA2AEX(_In_z_ LPCSTR psz, _In_ UINT nCodePage)
+    {
+        Init(psz);
+        UNREFERENCED_PARAMETER(nCodePage);
+    }
+
+    ~CA2AEX() throw()
+    {
+        if (m_psz != m_szBuffer)
+            free(m_psz);
+    }
+
+    _Ret_z_ operator LPSTR() const throw()
+    {
+        return m_psz;
+    }
+
+private:
+    // CA2AEX is not copyable
+    CA2AEX(_In_ const CA2AEX &) throw() = delete;
+    CA2AEX& operator=(_In_ const CA2AEX &) throw() = delete;
+
+    void Init(_In_z_ LPCSTR psz)
+    {
+        int cchMax = lstrlenA(psz) + 1;
+        if (cchMax <= t_nBufferLength)
+        {
+#ifdef _STRSAFE_H_INCLUDED_
+            StringCchCopyA(m_szBuffer, _countof(m_szBuffer), psz);
+#else
+            lstrcpynA(m_szBuffer, psz, _countof(m_szBuffer));
+#endif
+            m_psz = m_szBuffer;
+            return;
+        }
+
+        m_szBuffer[0] = 0;
+        m_psz = _strdup(psz);
+        if (!m_psz)
+            AtlThrow(E_OUTOFMEMORY);
+    }
+};
+
+template <int t_nBufferLength = 128>
+class CW2WEX
+{
+public:
+    LPWSTR m_psz;
+    wchar_t m_szBuffer[t_nBufferLength];
+
+    CW2WEX(_In_z_ LPCWSTR psz)
+    {
+        Init(psz);
+    }
+
+    CW2WEX(_In_z_ LPCWSTR psz, _In_ UINT nCodePage)
+    {
+        Init(psz);
+        UNREFERENCED_PARAMETER(nCodePage);
+    }
+
+    ~CW2WEX() throw()
+    {
+        if (m_psz != m_szBuffer)
+            free(m_psz);
+    }
+
+    _Ret_z_ operator LPWSTR() const throw()
+    {
+        return m_psz;
+    }
+
+private:
+    // CW2WEX is not copyable
+    CW2WEX(_In_ const CW2WEX&) throw() = delete;
+    CW2WEX& operator=(_In_ const CW2WEX&) throw() = delete;
+
+    void Init(_In_z_ LPCWSTR psz)
+    {
+        int cchMax = lstrlenW(psz);
+        if (cchMax <= t_nBufferLength)
+        {
+#ifdef _STRSAFE_H_INCLUDED_
+            StringCchCopyW(m_szBuffer, _countof(m_szBuffer), psz);
+#else
+            lstrcpynW(m_szBuffer, psz, _countof(m_szBuffer));
+#endif
+            m_psz = m_szBuffer;
+            return;
+        }
+
+        m_szBuffer[0] = 0;
+        m_psz = _wcsdup(psz);
+        if (!m_psz)
+            AtlThrow(E_OUTOFMEMORY);
+    }
+};
+
+template <int t_nBufferLength = 128>
+class CA2WEX
+{
+public:
+    LPWSTR m_psz;
+    wchar_t m_szBuffer[t_nBufferLength];
+
+    CA2WEX(_In_z_ LPCSTR psz)
+    {
+        Init(psz, CP_ACP);
+    }
+
+    CA2WEX(_In_z_ LPCSTR psz, _In_ UINT nCodePage)
+    {
+        Init(psz, nCodePage);
+    }
+
+    ~CA2WEX() throw()
+    {
+        if (m_psz != m_szBuffer)
+            free(m_psz);
+    }
+
+    _Ret_z_ operator LPWSTR() const throw()
+    {
+        return m_psz;
+    }
+
+private:
+    // CA2WEX is not copyable
+    CA2WEX(_In_ const CA2WEX&) throw() = delete;
+    CA2WEX& operator=(_In_ const CA2WEX&) throw() = delete;
+
+    void Init(_In_z_ LPCSTR psz, _In_ UINT nCodePage)
+    {
+        int cchMax = MultiByteToWideChar(nCodePage, 0, psz, -1, NULL, 0);
+        if (cchMax <= (int)_countof(m_szBuffer))
```

Comment:
I wonder what happens if `cchMax` happens to be equal to zero (due to whatever reason: e.g. empty string being converted etc.)...

Reply:
Added the tests of zero-length.